pilot/pkg/xds/discovery.go

```diff
@@ -252,7 +252,8 @@ func (s *DiscoveryServer) dropCacheForRequest(req *model.PushRequest) {
 
 // Push is called to push changes on config updates using ADS.
 func (s *DiscoveryServer) Push(req *model.PushRequest) {
-	if !req.Full {
+	// If it is headless service endpoint update, we need to re-init service registry to prevent generating stale nds.
+	if !req.Full && !req.Reason.Has(model.HeadlessEndpointUpdate) {
```

Comment:
Why not just make the code that makes a HeadlessEndpointUpdate set full=true?

Reply:
I thought about it, but this will restore the issue of introducing unnecessary push to lds/rds even not updated 
FYI https://github.com/istio/istio/issues/48207

Reply:
Hmm, good point. I do worry about making a `!Full` recompute pushcontext. 

Originally, `!Full` was very clear - no push context recompute, and only EDS pushes it. Then it became a bit blurrier, as other types started supporting `!Full`. But through all this time, it meant 'no push context recompute'.

Now this PR changes it, and `!Full` really has no meaning.

I also don't feel great about usage of `req.Reason` (which was originally intended only for metrics), but we use that all over the place so not as concerned... but I really don't want to totally break `!Full`.

Maybe we can just skip LDS/RDS if the `Reason is only model.HeadlessEndpointUpdate`....? Though that may not be great, either.

We could introduce a new `kind` for this type of change, like `DNSName` or something. The root of this seems to be we overload `kind: ServiceEntry`?

Reply:
Now we reply on `kind: ServiceEntry` to init serviceregistry, update it here, means we need to update the pushConext init too. This does not matter much.

Originally concern from me is i am not sure if recreate a new context for `! FULL`  has some side-effect, both this patch and your suggestion seem need to recreate the push context.

Reply:
I will double check it though. Especially the push status

```
	// Reset the status during the push.
	oldPushContext := s.globalPushContext()
	if oldPushContext != nil {
		oldPushContext.OnConfigChange()
		// Push the previous push Envoy metrics.
		envoyfilter.RecordMetrics()
	}
```